### ISSUE TITLE
internal/uidriver/glfw: Implement currentMonitorByOS on Linux or other UNIX

### ISSUE THREAD

**Issue by hajimehoshi** [hajimehoshi/ebiten#1119]
`currentMonitor` should not use the current window's position as much as possible, because this function can be accessed before the main loop, when the window position is not reliable.

Now, this always returns the primary monitor and the window always starts at the primary monitor, wherever the user is.

**Comment by beoran** [comment]
I will look into this and see if I can help with it.

**Comment by hajimehoshi** [comment]
@divVerent Hi, do you know how to get the current active monitor (without a window handler) via X protocol? Thanks,

**Comment by beoran** [comment]
I looked into it (ok, a bit late) , and I think we need to use the XRandR extension, and in particular the XRRGetOutputPrimary function. GLFW also uses this: https://github.com/glfw/glfw/blob/master/src/x11_monitor.c .

**Comment by hajimehoshi** [comment]
Thanks, but IIUC the primary display is a little different from the currently active display.

X protocol is already used at `ui_unix.go` so it's fine to use X protocol directly there.

**Comment by beoran** [comment]
Sorry to ask but what exactly is the "active" display? Is I have 2 monitors connected then what dermines this? The position of the mouse, or something else?

**Comment by beoran** [comment]
If it is simply the first connected screen (crtc) then we can use XRRGetCrtcInfo() and if available use the crts info at index 0 as the first "active" display.

https://github.com/liballeg/allegro5/blob/master/src/x/xrandr.c
https://stackoverflow.com/questions/13706078/xrandr-related-c-programming

Otherwise we might have to see where the mouse is with relation to the CRTC info, and perhaps the Display info as well.

**Comment by hajimehoshi** [comment]
The position of the mouse seems fine. (Or, the window might have an information which output it belongs to, but sometimes the window's position is not reliable)

**Comment by hajimehoshi** [comment]
OK I'try to implement the cursor-position way. This can be implemented only with GLFW APIs.

**Comment by hajimehoshi** [comment]
I'm not sure the general rule which monitor should be used when launching a GUI application, but detecting the cursor position seems general.

**Comment by beoran** [comment]
Yes, that's also how we seem to do it for Allegro: 
See _al_xsys_mheadx_get_default_adapter in
https://github.com/liballeg/allegro5/blob/4aa54e6c994af21bc63d8b593673ab3df62390f8/src/x/xfullscreen.c

**Comment by hajimehoshi** [comment]
Hmm, when a cursor and a terminal is in different monitors, and you press the enter key to launch something, the app should be opened on the terminal side.

**Comment by hajimehoshi** [comment]
> 
https://github.com/liballeg/allegro5/blob/4aa54e6c994af21bc63d8b593673ab3df62390f8/src/x/xfullscreen.c

`XGetInputFocus` seems the key, thanks

**Comment by beoran** [comment]
Yes, but as you can see in tha tcode we confirm it with a check with the mouse cursor position.

**Comment by hajimehoshi** [comment]
https://pkg.go.dev/github.com/jezek/xgb@v0.0.0-20210312150743-0e0f116e1240/xproto#GetWindowAttributesReply `Screen` doesn't exist in this library 😢

**Comment by hajimehoshi** [comment]
OK probably I can do it by myself https://gitlab.freedesktop.org/xorg/lib/libx11/-/blob/master/src/GetWAttrs.c#L130-139

**Comment by beoran** [comment]
It is also missing in xproto, so I reported it: https://gitlab.freedesktop.org/xorg/proto/xcbproto/-/issues/21.
Meanwhile we I guess might have to think of a different way...

**Comment by beoran** [comment]
You are right, actually _XGetWindowAttributes doesn't get this from X11 but determines it itself by looping over the available screens in the display. I'll close that issue again.

**Comment by beoran** [comment]
You may have to make a few X requests, actually, to get the root, including https://pkg.go.dev/github.com/jezek/xgb@v0.0.0-20210312150743-0e0f116e1240/xproto#GetGeometry .

**Comment by divVerent** [comment]
I think it depends a lot on what you _mean_ by current monitor.

E.g. which monitor a fullscreen window ends up on is not clearly defined (_NET_WM_STATE_FULLSCREEN does not let you select a specific monitor, and neither does the spec say which one to use).

However, and this may be very relevant if this is for ScreenSizeInFullScreen() - GLFW actually can pass a second hint: _NET_WM_FULLSCREEN_MONITORS to do request a specific one. GLFW does this when window->monitor is set.

My personal preference is "same monitor the mouse cursor is on", but then of course Ebiten will also have to make sure to actually create its fullscreen windows on that screen. However, IIRC this is a preference that may differ for people and window managers may have configuration for this. IIRC the only "proper" way to say "which screen should a new window appear on" is actually creating the window and checking what happened. And indeed, "same monitor as the currently focused window" is just as valid, and querying _NET_WM_ACTIVE_WINDOW will usually be a better way to do this, although not all window managers support this (so you'll need the XGetInputFocus-and-then-walk-upwards-towards-root way anyway).

Now, checking in Ebiten source what we actually use currentMonitorByOS _for_... it appears to be used only to find the monitor the current window is _on_. So this shouldn't use mouse cursor or anything, but "some coordinate" of the window, _if_ it's already been fully created (heuristics like focus or mouse position however make sense _before_ we have a real window).

X11 does in general not know what a monitor is, and e.g. does not prevent windows from spanning multiple monitors. I would even say that getMonitorFromPosition is _the_ correct way on X11. Multiple-monitor support is usually implemented at window manager level, using XRandR to query information about the monitors and to then limit window placement to prevent them from spanning monitors and to make sure maximizing fills a monitor entirely etc.

**Comment by hajimehoshi** [comment]
>  IIRC the only "proper" way to say "which screen should a new window appear on" is actually creating the window and checking what happened. And indeed, "same monitor as the currently focused window" is just as valid, and querying _NET_WM_ACTIVE_WINDOW will usually be a better way to do this, although not all window managers support this (so you'll need the XGetInputFocus-and-then-walk-upwards-towards-root way anyway).

Thanks. Yes, what I want is to know which monitor a new window is created. Ebiten has a transparent window before the main loop, so getting the root window from the window might work (`GetGeometry`) but I'm not sure. `_NET_WM_ACTIVE_WINDOW` seems also the way to go. Even this is not supported, that's fine, we can use the primary monitor as a fallback.

**Comment by hajimehoshi** [comment]
> (so you'll need the XGetInputFocus-and-then-walk-upwards-towards-root way anyway).

My understanding is there is one root window and the root window includes all the monitors. Actually, `videoModeScaleUncached` now gets the default root window and iterates the CRTC information. Is this correct, or does this depend on window managers?

**Comment by divVerent** [comment]
The times when people on X11 had multiple root windows are over. That is
from the early beginnings of multiscreen, when one ran a separate window
manager on each monitor using displays like :0.0, :0.1 etc. - and even for
that, Ebiten basically can assume there is just one monitor, as it will not
even have access to anything else.

Let's not worry about that. After that Xinerama and finally XRandR came
along to do it in a better way, actually allowing interaction and moving
windows between screens. And yes, with those you have exactly one root
window spanning all monitors, and it is the window manager's job to ensure
non-stupid window placement.

The walking upwards is needed for another reason - XGetInputFocus may
return a subwindow and you probably want the closest child to root, so you
walk up to root then take the last non root window on the path.

**Comment by divVerent** [comment]
On Wed, Sep 22, 2021, 23:19 Hajime Hoshi ***@***.***> wrote:

> IIRC the only "proper" way to say "which screen should a new window appear
> on" is actually creating the window and checking what happened. And indeed,
> "same monitor as the currently focused window" is just as valid, and
> querying _NET_WM_ACTIVE_WINDOW will usually be a better way to do this,
> although not all window managers support this (so you'll need the
> XGetInputFocus-and-then-walk-upwards-towards-root way anyway).
>
> Thanks. Yes, what I want is to know which monitor a new window is created.
> Ebiten has a transparent window before the main loop, so getting the root
> window from the window might work (GetGeometry) but I'm not sure.
> _NET_WM_ACTIVE_WINDOW seems also the way to go. Even this is not
> supported, that's fine, we can use the primary monitor as a fallback.
>
I guess that is OK, but even that is only approximate.

Especially tiling window managers (and Wayland tiling compositors like
Sway) allow defining in a config file which window gets to which screen
regardless of mouse position or focus. Usually defined based on window
class names or title.

So basically, for windowed mode one cannot predict at all. For fullscreen
mode, as I wrote above, you kinda can - you can request a specific monitor,
and then assume that is the one you will get.

And yes, this is kinda bad if you want to use the device scale factor of
the monitor before creating a window so you know which size to request -
basically a chicken/egg problem. The good news is that X11 only supports
global scale (via Xft.dpi); the only thing that varies per monitor is the
final scale from framebuffer to screen, which you set using XRandR, but you
do not need that to compute a desired window size.

Of course, this is something Wayland solves better, by allowing XRandR-ish
scaling but unlike X11, allowing native-DPI surfaces on it. But in
particular XWayland does not support that, so anyone using this will look
like LoDPI to X11 apps.

**Comment by hajimehoshi** [comment]
Now I've succeeded to get a correct position from an initial window, I can de-prioritize this task. (A hidden window just after creation always had (0, 0) position, and this was the problem. After showing and hiding it once, it has a correct position. Ha! See https://github.com/hajimehoshi/ebiten/commit/77664ff0575c8f51e1ba5cd3604a57808fbf95f4 for more details)

> So basically, for windowed mode one cannot predict at all.

Hmm ok. Thank you for elaborating. Apparently there is no silver bullet there.

Currently Ebiten has a fallback to detect a monitor from the window position, and this might be already enough.

**Comment by hajimehoshi** [comment]
Integrated into #1982

**Comment by hajimehoshi** [comment]
Oops this is different from `initialMonitorByOS`
